src/adb.js

```diff
@@ -1 +1,353 @@
+		return conn
+			.exec('host:transport:' + deviceId)
+			.then(data => {
+				return conn.exec('shell:' + cmd.replace(/^shell\:/, ''), { bufferUntilClose: true, noLength: true });
+			})
+			.then(result => result)
+			.catch(err => {
+				console.log('adb shell: ', err);
+			});
+	}
+
+	/**
+	 * Attempts to find the adb executable, then start the adb server.
+	 * @returns {Promise}
+	 */
+	startServer() {
+		return sdk
+			.detect(this.config)
+			.then(result => {
+				return util.run(result.sdk.executables.adb, ['start-server']);
+			}).catch(err => {
+				console.log('start server', err);
+			});
+	}
+
+	/**
+	 * Attempts to find the adb executable, then stop the adb server.
+	 * @returns {Promise}
+	 */
+	stopServer() {
+		return sdk
+			.detect(this.config)
+			.then(result => {
+				return util.run(result.sdk.executables.adb, ['kill-server']);
+			})
+			.catch(err => {
+				console.log('stop server', err);
+			});
+	}
+
+	/**
+	 * Forwards the specified device/emulator's socket connections to the destination.
+	 * @param {String} deviceId - The id of the device or emulator
+	 * @param {String} src - The source port in the format "tcp:<port>"
+	 * @param {String} dest - The destination port in the format "tcp:<port>" or "jdwp:<pid>"
+	 * @returns {Promise}
+	 */
+	forward(deviceId, src, dest) {
+		return sdk
+			.detect(this.config)
+			.then(result => {
+				return util.run(result.sdk.executables.adb, ['-s', deviceId, 'forward', src, dest]);
+			})
+			.catch(err => {
+				console.log('stop server', err);
+			});
+	}
+
+	/**
+	 * Pushes a single file to a device or emulator.
+	 * @param {String} deviceId - The id of the device or emulator
+	 * @param {String} src - The source file to copy to the device
+	 * @param {String} dest - The destination to write the file
+	 * @returns {Promise}
+	 */
+	push(deviceId, src, dest) {
+		src = path.resolve(src);
+		if (!fs.existsSync(src)) {
+			return Promise.reject('Source file "%s" does not exist', src);
+		}
+
+		return sdk
+			.detect(this.config)
+			.then(result => {
+				return util.run(result.sdk.executables.adb, ['-s', deviceId, 'push', src, dest]);
+			})
+			.catch(err => {
+				console.log('stop server', err);
+			});
+	}
+
+	/**
+	 * Pulls a single file from a device or emulator.
+	 * @param {String} deviceId - The id of the device or emulator
+	 * @param {String} src - The source file to copy from the device
+	 * @param {String} dest - The destination to write the file
+	 * @returns {Promise}
+	 */
+	pull(deviceId, src, dest) {
+		dest = path.resolve(dest);
+		const destDir = path.dirname(dest);
+
+		fs.existsSync(destDir) || wrench.mkdirSyncRecursive(destDir);
+
+		return sdk
+			.detect(this.config)
+			.then(result => {
+				return util.run(result.sdk.executables.adb, ['-s', deviceId, 'pull', src, dest]);
+			})
+			.catch(err => {
+				console.log('stop server', err);
+			});
+	}
+
+	/**
+	 * Returns the pid of the specified app and device/emulator, if running.
```

Comment:
"specified app and device/emulator" should be "specified app for the specified device/emulator".